modules/renter.go

```diff
@@ -50,12 +54,92 @@ type Allowance struct {
-	Received    uint64    `json:"received"`
-	StartTime   time.Time `json:"starttime"`
-	Error       string    `json:"error"`
+	SiaPath     string         `json:"siapath"`
+	Destination DownloadWriter `json:"destination"`
+	Filesize    uint64         `json:"filesize"`
+	Received    uint64         `json:"received"`
+	StartTime   time.Time      `json:"starttime"`
+	Error       string         `json:"error"`
+}
+
+// DownloadWriter provides an interface which all output writers have to implement.
+type DownloadWriter interface {
+	WriteAt(b []byte, off int64) (int, error)
+}
+
+// DownloadFileWriter is a file-backed implementation of DownloadWriter.
+type DownloadFileWriter struct {
+	f        *os.File
+	Location string
+	offset   uint64
+}
+
+// NewDownloadFileWriter creates a new instance of a DownloadWriter backed by the file named.
+func NewDownloadFileWriter(fname string, offset, length uint64) *DownloadFileWriter {
+	l, _ := os.OpenFile(fname, os.O_CREATE|os.O_WRONLY, defaultFilePerm)
+	return &DownloadFileWriter{
+		f:        l,
+		Location: fname,
+		offset:   offset,
+	}
+}
+
+// WriteAt writes the passed bytes at the specified offset.
+func (dw *DownloadFileWriter) WriteAt(b []byte, off int64) (int, error) {
+	fileOffset := off - int64(dw.offset)
+
+	r, err := dw.f.WriteAt(b, fileOffset)
+	if err != nil {
+		build.ExtendErr("unable to write to download destination", err)
+	}
+	dw.f.Sync()
+
+	return r, err
+}
+
+// DownloadHttpWriter is a http response writer-backed implementation of DownloadWriter.
+// The writer writes all content that is written to the current `offset` directly to the ResponseWriter,
+// and buffers all content that is written at other offsets.
+// After every write to the ResponseWriter the `offset` and `length` fields are updated, and buffer content written until
+type DownloadHttpWriter struct {
+	w              http.ResponseWriter
+	offset         int         // The index in the original file of the last byte written to the response writer.
+	firstByteIndex int         // The index of the first byte in the original file.
+	length         int         // The total size of the slice to be written.
+	buffer         []byte      // Buffer used for storing the chunks until download finished.
+	offsets        map[int]int // Map from offset to length of chunk.
+}
+
+// NewDownloadHttpWriter creates a new instance of http.ResponseWriter backed DownloadWriter.
+func NewDownloadHttpWriter(w http.ResponseWriter, offset, length uint64) *DownloadHttpWriter {
+	return &DownloadHttpWriter{
+		w:              w,
+		offset:         int(offset),
+		firstByteIndex: int(offset),
+		length:         int(length),
+		buffer:         make([]byte, length),
+		offsets:        make(map[int]int),
+	}
+}
+
+// WriteAt buffers parts of the file until the entire file can be
+// flushed to the client. Returns the number of bytes written or an error.
+func (dw *DownloadHttpWriter) WriteAt(b []byte, off int64) (int, error) {
```

Comment:
**Re: Buffer**: I've got a pretty straight forward fix for the buffering problem, should get it out by EOD. I had a working version of this in previous commits but chose to take it out because the tests simulate a too-perfect environment (e.g. chunks always arrive in order because they are scheduled in order).

**Re: Download overlap**: The actual downloading logic hasn't changed, just the destination of the data. The initial layout of the section download was completely separate (see initial reviews), which introduced many corner cases, made the code unwieldy and prone to logic duplication.
@lukechampine could you elaborate on what you mean by streaming vs parallel optimizations? Downloads are currently performed in parallel and order is only introduced by the order in which chunks are scheduled for download (which is currently linear).
If a `stream` flag were to be introduced to speed up streaming downloads it could be used with file-based and streaming downloads alike needing changes only to be made in one place.